manager/controllers/app/storage.go

```diff
@@ -6,168 +6,95 @@ package app
-	if resource.Status.AssetPrefixPerDataset == nil {
-		return ""
-	}
-	elem, ok := resource.Status.AssetPrefixPerDataset[id]
-	if !ok {
-		return ""
-	}
-	return elem
+	r.datasets = append(r.datasets, dataset)
+	return nil
 }
 
-// AddPrefix adds a prefix generated for the given data set
-func AddPrefix(resource *app.M4DBucket, id string, prefix string) {
-	if resource.Status.AssetPrefixPerDataset == nil {
-		resource.Status.AssetPrefixPerDataset = make(map[string]string)
+// AllocateBucket allocates a bucket in the relevant geo
+func AllocateBucket(c client.Client, log logr.Logger, owner types.NamespacedName, id string, geo string) (*comv1alpha1.Dataset, error) {
+	ctx := context.Background()
+	log.Info("Searching for a storage account matching the geography " + geo)
+	var accountList app.M4DStorageAccountList
+	if err := c.List(ctx, &accountList); err != nil {
+		log.Info(err.Error())
+		return nil, err
 	}
-	resource.Status.AssetPrefixPerDataset[id] = prefix
-}
-
-// GetOrCreatePrefix creates a new prefix for the given data set if none exists, based on the given name
-func GetOrCreatePrefix(resource *app.M4DBucket, id string, name string) string {
-	prefix := GetPrefix(resource, id)
-	if prefix != "" {
-		return prefix
+	for _, account := range accountList.Items {
+		utils.PrintStructure(account, log, "Account ")
+		if account.Spec.Region != geo {
+			continue
+		}
+		genName := generateDatasetName(owner, id)
+		values := map[string]string{
+			"type":        "COS",
+			"secret-name": account.Spec.SecretRef,
+			"endpoint":    account.Spec.Endpoint,
+			"bucket":      genName,
+			"provision":   "true"}
+		dataset := &comv1alpha1.Dataset{ObjectMeta: metav1.ObjectMeta{
+			Name:      genName,
+			Namespace: utils.GetSystemNamespace(),
+			Labels:    map[string]string{"m4d.ibm.com/owner": owner.Namespace + "." + owner.Name},
+		},
+			Spec: comv1alpha1.DatasetSpec{Local: values},
+		}
+		return dataset, nil
 	}
-	prefix = name + utils.Hash(name, 10)
-	AddPrefix(resource, id, prefix)
-	return prefix
+	return nil, errors.New("Could not allocate a bucket in " + geo)
 }
 
-// RemovePrefix removes a prefix generated for the given data set
-func RemovePrefix(resource *app.M4DBucket, id string) {
-	delete(resource.Status.AssetPrefixPerDataset, id)
-}
-
-// IsBucketAvailable checks whether the bucket is available for use
-func IsBucketAvailable(resource *app.M4DBucket, owner types.NamespacedName, id string, canShare bool) bool {
-	// is owned by any resource?
-	if !HasOwners(resource) {
-		return true
-	}
-	// is owned by this resource only?
-	if HasOwner(resource, owner) && len(resource.Status.Owners) == 1 {
-		return true
-	}
-	// sharing available for the given data set
-	if canShare && GetPrefix(resource, id) != "" {
-		return true
-	}
-	return false
+func generateDatasetName(owner types.NamespacedName, id string) string {
+	name := owner.Name + "-" + owner.Namespace + "-" + utils.Hash(id, 20)
```

Comment:
Should perhaps the source data set name be included in the name?

Reply:
OK.